File: examples/python/bufr_subset.py
# Copyright 2005-2016 ECMWF.
#
# This software is licensed under the terms of the Apache Licence Version 2.0
# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
#
# In applying this licence, ECMWF does not waive the privileges and immunities
# granted to it by virtue of its status as an intergovernmental organisation
# nor does it submit to any jurisdiction.

#
# Python implementation: bufr_subset
#
# Description: how to read data values from a given subset of a BUFR message.
#
#

import traceback
import sys

from eccodes import *

INPUT = '../../data/bufr/synop_multi_subset.bufr'
VERBOSE = 1  # verbose error reporting


def example():

    # open bufr file
    f = open(INPUT)

    cnt = 0

    # loop for the messages in the file
    while 1:
        # get handle for message
        bufr = codes_bufr_new_from_file(f)
        if bufr is None:
            break

        print "message: %s" % cnt

        # we need to instruct ecCodes to expand all the descriptors
        # i.e. unpack the data values
        codes_set(bufr, 'unpack', 1)

        # find out the number of subsets
        key = 'numberOfSubsets'
        numberOfSubsets = codes_get(bufr, 'numberOfSubsets')
        print ' %s: %d' % (key, numberOfSubsets)

        # loop over the subsets
        for i in range(1, numberOfSubsets + 1):

            # read and print some data values

            key = '/subsetNumber=%d/blockNumber' % i
            print key
            val = codes_get_long(bufr, key)
            print '  %s= %d' % (key, val)

            key = '/subsetNumber=%d/stationNumber' % i
            val = codes_get_long(bufr, key)
            print '  %s: %d' % (key, val)

        cnt += 1

        # delete handle
        codes_release(bufr)

    # close the file
    f.close()


def main():
    try:
        example()
    except CodesInternalError as err:
        if VERBOSE:
            traceback.print_exc(file=sys.stderr)
        else:
            sys.stderr.write(err.msg + '\n')

        return 1

if __name__ == "__main__":
    sys.exit(main())

File: examples/C/bufr_subset.c
/*
 * Copyright 2005-2018 ECMWF.
 *
 * This software is licensed under the terms of the Apache Licence Version 2.0
 * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * In applying this licence, ECMWF does not waive the privileges and immunities granted to it by
 * virtue of its status as an intergovernmental organisation nor does it submit to any jurisdiction.
 */

/*
 * C Implementation: bufr_subset
 *
 * Description: how to read data values from a given subset of a BUFR message.
 *
 */

#include "eccodes.h"

int main(int argc,char* argv[])
{
    char key[200]={0,};
    FILE* in = NULL;

    /* message handle. Required in all the eccodes calls acting on a message.*/
    codes_handle* h=NULL;

    long numberOfSubsets=0;
    long longVal;
    double doubleVal;
    size_t stringLen;
    char stringVal[100]={0,};
    int i,err=0;
    int cnt=0;
    const char* infile = "../../data/bufr/synop_multi_subset.bufr";

    in=fopen(infile,"r");
    if (!in) {
        printf("ERROR: unable to open file %s\n", infile);
        return 1;
    }

    /* loop over the messages in the bufr file */
    while ((h = codes_handle_new_from_file(NULL,in,PRODUCT_BUFR,&err)) != NULL || err != CODES_SUCCESS)
    {
        if (h == NULL) {
            printf("Error: unable to create handle for message %d\n",cnt);
            cnt++;
            continue;
        }

        printf("message: %d\n",cnt);

        /* we need to instruct ecCodes to expand all the descriptors i.e. unpack the data values */
        CODES_CHECK(codes_set_long(h,"unpack",1),0);

        /* find out the number of subsets */
        CODES_CHECK(codes_get_long(h,"numberOfSubsets",&numberOfSubsets),0);
        printf("  numberOfSubsets: %ld\n",numberOfSubsets);

        /* loop over the subsets */
        for(i=1; i <= numberOfSubsets; i++)
        {
            sprintf(key,"/subsetNumber=%d/blockNumber",i);

            printf("  subsetNumber=%d",i);
            /* read and print some data values */
            CODES_CHECK(codes_get_long(h,key,&longVal),0);
            printf("  blockNumber=%ld",longVal);

            sprintf(key,"/subsetNumber=%d/stationNumber",i);
            CODES_CHECK(codes_get_long(h,key,&longVal),0);
            printf("  stationNumber=%ld",longVal);

            sprintf(key,"/subsetNumber=%d/stationOrSiteName",i);
            stringLen=100;
            CODES_CHECK(codes_get_string(h,key,stringVal,&stringLen),0);
            printf("  stationOrSiteName=\"%s\"",stringVal);

            sprintf(key,"/subsetNumber=%d/airTemperature",i);
            CODES_CHECK(codes_get_double(h,key,&doubleVal),0);
            printf("  airTemperature=%g\n",doubleVal);

        }

        /* delete handle */
        codes_handle_delete(h);

        cnt++;
    }

    fclose(in);
    return 0;
}
